### ISSUE TITLE
[DAY 36] Lightweight models & The flip-side of the coin: on-device AI & Optimization

### ISSUE THREAD

**Issue by changwoomon** [boostcamp-ai-tech-4/peer-session#103]
#### `원딜` @changwoomon 

* 캐글 책 베타리더 당첨!

#### `후미` @opijae 

* 롤 골드 달성!

#### `샐리` @bsm8734 

* 친구분들과 사진관가서 사진 찍으셨습니다.

#### `히스` @Heeseok-Jeong 

* 지난주 강의 복습, 백준 쉬운 문제 풀기, 서핑 하셨습니다.

#### `펭귄` @CoodingPenguin 

* 유튜브(브레이브 걸스 영상) 시청, 백화점 쇼핑 하셨습니다.

---

#### 2) 변환 후 모델의 size와 parameter의 개수는 어떻게 될까?
- 보통 size와 parameter 개수는 비례 → size를 줄이면 parameter 개수도 당연히 줄지 않을까요?
- PyTorch(.pth) : 233MB
- onnx : 233MB
- TF(.pb) : 233MB
- Keras(.h5) : 233MB
- **TF Lite(.tflite) : 58MB**
- **TensorRT(.trt) : 466MB**
  - 왜 TensorRT로 바꾼 모델이 크기가 가장 클까요?
    - [TensorRT](https://m.blog.naver.com/PostView.nhn?blogId=sogangori&logNo=220985061392&proxyReferer=https:%2F%2Fwww.google.com%2F)
    - tensorRT 는 tflite 와 같이 on-device 의 목적이 아닌 NVIDIA GPU 에서 inference 속도를 높여주는 것이 목적이기 때문에 변수를 NVIDIA GPU 에 최적화시키는 과정에서 용량은 커지지만 inference 속도는 빨라진다고 생각했습니다.
- onnx → pth : 351MB
  - pth → onnx → pth 과정으로 진행되었는데 왜 용량이 더 커졌을까요?
    - https://github.com/boostcamp-ai-tech-4/peer-session/issues/102
#### 3) 이 중에서 가장 가벼운 모델은 어떤 것일까? (슬라이드 p.41)
- TF Lite(.tflite) : 58MB
#### 4) 어떤 변환은 가역적이고 어떤 변환은 비가역적일까?
- PyTorch ↔ onnx → TF / Keras / TF Lite / TensorRT
<br>

#### 1) 이번 강의는 lightweight modeling과 어떤 관계가 있을까? (이 강의를 만든 목적은 무엇일까?)
- 비즈니스를 위해서 on-device로 옮기는 과정이 필요한데 이를 위해 lightweight 모델을 만들어서 옮겨야 한다.
#### 2) 고객의 어떤 문제를 내가 해결할 수 있을까? + 3) on-device solution의 다른 아이디어는 어떤 것이 있을까?
- [차세대 딥러닝 기술 ‘온 디바이스(On-Device) AI’란?](https://www.samsungsemiconstory.com/2184)
- 스마트폰으로 마스크 썼는지 구별
- 공사장, 야외에서 이용
- GPS가 안되는 곳에서 내가 어디있는지 
#### 4) Memory 제약이 얼만큼이면 vgg16 모델이 작동을 못하는지 기록해보자
<br>

#### 1) 이번 강의는 lightweight modeling과 어떤 관계가 있을까? (이 강의를 만든 목적은 무엇일까?)
- 도메인에 따라 cost가 달라진다 → 각각 cost에 따라 모델을 경량화 할 수 있다.
#### 2) euclidean / hamming / cosine distance의 차이는? 언제 어떤 거리를 사용해야할까?
- [Euclidean, Angular(Cosine), Hamming Distance 차이](https://m.blog.naver.com/boy_1206/221806208758)
- [4 Types of Distance Metrics in Machine Learning](https://www.analyticsvidhya.com/blog/2020/02/4-types-of-distance-metrics-in-machine-learning/)
- [4 Distance Measures for Machine Learning](https://machinelearningmastery.com/distance-measures-for-machine-learning/)
- **euclidean distance**
  - [유클리드 거리](https://ko.wikipedia.org/wiki/%EC%9C%A0%ED%81%B4%EB%A6%AC%EB%93%9C_%EA%B1%B0%EB%A6%AC)
  - 좌표계
- **hamming distance**
  - [해밍 거리](https://ko.wikipedia.org/wiki/%ED%95%B4%EB%B0%8D_%EA%B1%B0%EB%A6%AC)
  - [Hamming Distance Metric Learning
](https://papers.nips.cc/paper/2012/file/59b90e1005a220e2ebc542eb9d950b1e-Paper.pdf)
  - 비트, 문자열
  - 왜 NLP에서 많이 안쓸까요?
    - NLP에서는 character 별로 비교하기보다는 word 단위로 비교하기 때문에..?
    - word → Embedding → vector가 되어 cosine 유사도를 사용할 것이다.
- **cosine distance**
  - [코사인 유사도](https://ko.wikipedia.org/wiki/%EC%BD%94%EC%82%AC%EC%9D%B8_%EC%9C%A0%EC%82%AC%EB%8F%84)
  - 벡터
#### 3) decision problem과 optimization problem의 차이를 적어보자
- decision problem
  - target에만 도달하면 된다. cost, loss가 작아지기만 하면 decision problem을 푸는 것이다.
- optimization problem 
  - decision problem을 반복하면서 cost, loss를 줄이는 방향으로 문제를 푼다.
#### 4) backpropagation 식이 왜 이렇게 될까?
```python
    def sigmoid_derivative(self, x):
        return x * (1.0 - x)
```
  <img src="https://user-images.githubusercontent.com/59998179/111140913-3b4f3e80-85c6-11eb-90bf-b69a9ed29d6a.png" width=500>
  <img src="https://user-images.githubusercontent.com/59998179/111141066-6c2f7380-85c6-11eb-91ff-064aaee93774.png" width=600>
